Mathlib/LinearAlgebra/Multilinear/Basic.lean

```diff
@@ -760,8 +763,204 @@ theorem domDomCongr_eq_iff (σ : ι₁ ≃ ι₂) (f g : MultilinearMap R (fun _
 
 end
 
+
+section Derivative
+
+/-! The goal of this part is, for `f` a multilinear map in finitely many variables indexed by `ι`,
+to develop `f(x + y)` as a finite formal multilinear series in `y`. The term of degree `n` of
+this formal multilinear series will be a sum over finsets `s` of `ι` of cardinality `n` of a
+multilinear map indexed by `s`, evaluated at `fun (_ : s) => y`; we use a linear order on
+`ι` to identify all such `s` to `Fin n`. We also give a more direct definition of the term
+of degree `1` in `f.linearDeriv`, which doesn't require a linear order on `ι`, and prove the
+equivalence of the two definitions.
+
+We start by introducing the individual terms indexed by finsets of `ι`; actually, for the
+definition, we can consider any set of `ι` and we don't need to take `ι` finite.-/
+
+/-- Given a multilinear map `f` on `(i : ι) → M i`, an element `x` of `(i : ι) → M i` and s
+set `s` of `ι`, construct a multilinear map on `s → ((i : ι) → Mi)` whose value at `z`
+is `f` evaluated at the vector with `i`th coordinate `(z i) i` if `i ∈ s` and `x i` otherwise.-/
+def toMultilinearMap_set [DecidableEq ι] (f : MultilinearMap R M₁ M₂) (x : (i : ι) → M₁ i)
+    (s : Set ι) [(i : ι) → Decidable (i ∈ s)] :
+    MultilinearMap R (fun (_ : s) => (i : ι) → M₁ i) M₂ where
```

Comment:
Are you sure you don't want
```suggestion
    MultilinearMap R (fun (i : s) => M₁ i) M₂ where
```
which avoids the "diagonalization" of `z`, replacing `z ⟨i, h⟩ i` with `z ⟨i, h⟩` below?

Reply:
I eventually want a [FormalMultilinearSeries](https://leanprover-community.github.io/mathlib4_docs/Mathlib/Analysis/Calculus/FormalMultilinearSeries.html#FormalMultilinearSeries) that sums to `f`, when the multilinear map `f` is also continuous. As `f` is a map from `(i : ι) → M₁` to `M₂`, the nth term of the FormalMultilinearSeries must be a multilinear map on `Fin n →((i : ι) → M₁)` (and then we evaluate it on the diagonal when we sum). This was the most direct way I found to do that.

I certainly do *not* want to restrict `f` to the components in `s`, and I do want the result to depend on `x`. I wonder if I might do what I want to do in another way, which would avoid the ugly definition that you point out and avoid the choice of the linear order. I need to write stuff down on paper.

Reply:
So I don't need to choose an order on `ι`, just an equivalence of each `s` of cardinality `n` with `Fin n`. Which choice is ugliest is up for debate. The mathematical point hiding behind that is that the nth derivative is not unique in positive characteristic. (In characteristic 0, I could just take the symmetrization of the thing I defined and divide by `n!`.)
As for the original question, yes, I could start by defining a multilinear map on `(i : s) → M₁ i` and build the nth derivative from that. I guess the auxiliary construction could having an independent interest, so let me try.

Reply:
Yes, I think my claim is mainly "you could factor this through the definition I propose", and that the factored definition could be called `domDomRestrict`

Reply:
Working on it right now.

Reply:
Done. Now let's see what CI makes of this.